### PR TITLE
Ensure ObjectDescription signature nodes include line numbers

### PR DESCRIPTION
**Pull request by jbms**
### Feature or Bugfix

- Bugfix

### Purpose
Previously, if any options were specified to the directive, line
numbers were not properly included in the signature nodes.

### CODE REVIEW

**Comment by tk0miya** [comment]
Could you explain your problem, please? Does it happen only with the directives implemented via ObjectDescription?

**Comment by jbms** [comment]
You can get a sense of the problem from the test case --- basically in:

```rst
.. py:function foo(bar: X)
   :noindexentry:

   Description
```

The signature line won't have correct source line numbers, which means, for example, that if the `X` type can't be resolved the warning message won't show correct line numbers.

Due to the way docutils parses directives, it only happens if there are options to the directive.  At first I thought it may be a bug in docutils, but I think it is actually the responsibility of the directive itself to sort out line numbers anyway, since only the directive knows which source line is actually the source of a given node that it produces.

For the signature, sphinx already attempts to set the line number to match the first line of the directive here:
https://github.com/sphinx-doc/sphinx/blob/b12b39db05d605cfb946f84afd27e0f6cf3c9e13/sphinx/directives/__init__.py#L175

However, in many cases the line numbers are instead obtained from `self.state.document.current_line` as child nodes are added, so setting the source info on the root signature node is not sufficient.

**Comment by jbms** [comment]
@tk0miya Is there anything left to do on this?

**Comment by jbms** [comment]
@tk0miya It would be great to get this fix in.  Do you think you might have chance to review it again soon?

File: tests/test_domain_py.py [hunk 1]
```diff
@@ -1351,3 +1352,16 @@ def test_warn_missing_reference(app, status, warning):
...
             in warning.getvalue())
+
+
+@pytest.mark.sphinx(confoverrides={'nitpicky': True})
+@pytest.mark.parametrize('include_options', (True, False))
+def test_signature_line_number(app, include_options):
+    text = (".. py:function:: foo(bar : string)\n" +
+            ("   :noindexentry:\n" if include_options else ""))
+    doc = restructuredtext.parse(app, text)
+    xrefs = list(doc.traverse(condition=addnodes.pending_xref))
```
Comment (AA-Turner) [inline tests/test_domain_py.py]: This should be `.findall()`

A

**Review by AA-Turner — COMMENTED** [review]
I should be able to review over the weekend.

A

File: tests/test_domain_py.py (hunk 1 above)
Comment (jbms) [reply tests/test_domain_py.py]: Thanks --- fixed.  I had thought that `traverse` already returned an iterator, but I see that `traverse(...)` is equivalent to `list(findall(...))`

File: sphinx/directives/__init__.py [hunk 2]
```diff
@@ -157,6 +157,15 @@ def run(self) -> List[Node]:
...
         node.document = self.state.document
+        source, line = self.get_source_info()
+        # If any options were specified to the directive,
+        # `self.state.document.current_line` will at this point be set to
+        # `None`.  To ensure nodes created as part of the signature have a line
+        # number set, set the document's line number correctly.
+        #
+        # Note that we need to subtract one from the line number since
+        # `note_source` uses 0-based line numbers.
+        self.state.document.note_source(source, line - 1 if line is not None else None)
```
Comment (AA-Turner) [inline sphinx/directives/__init__.py]: The ternary within the function call is hard to parse:
```suggestion
        if line is not None:
            line -= 1
        self.state.document.note_source(source, line)
```

**Review by AA-Turner — APPROVED** [review]

File: sphinx/directives/__init__.py [hunk 3]
```diff
@@ -157,6 +157,17 @@ def run(self) -> List[Node]:
 
         node = addnodes.desc()
         node.document = self.state.document
+        source, line = self.get_source_info()
+        # If any options were specified to the directive,
+        # `self.state.document.current_line` will at this point be set to
```
Comment (AA-Turner) [inline sphinx/directives/__init__.py]:
```suggestion
        # self.state.document.current_line will at this point be set to
```

File: sphinx/directives/__init__.py [hunk 4]
```diff
@@ -157,6 +157,17 @@ def run(self) -> List[Node]:
...
+        # `None`.  To ensure nodes created as part of the signature have a line
```
Comment (AA-Turner) [inline sphinx/directives/__init__.py]:
```suggestion
        # None.  To ensure nodes created as part of the signature have a line
```

File: sphinx/directives/__init__.py [hunk 5]
```diff
@@ -157,6 +157,17 @@ def run(self) -> List[Node]:
...
+        # number set, set the document's line number correctly.
+        #
+        # Note that we need to subtract one from the line number since
+        # `note_source` uses 0-based line numbers.
```
Comment (AA-Turner) [inline sphinx/directives/__init__.py]:
```suggestion
        # note_source uses 0-based line numbers.
```

**Comment by AA-Turner** [comment]
The markup within comments was incorrect (single rather than double backtick), but I just removed as it is within comments.

A
